optas/solver.py

```diff
@@ -100,17 +100,99 @@ def _solve(self):
 
     def solve(self):
         """Solve the optimization problem."""
-        return self.opt.decision_variables.vec2dict(self._solve())
+        solution = self.opt.decision_variables.vec2dict(self._solve())
+        # Add full model state to the solution dictionary
+        for model in self.opt.models:
+            for d in model.time_derivs:
+                n_s = model.state_name(d)
+                if isinstance(model, RobotModel):
+                    n_s_x = model.state_optimized_name(d)
+                    n_s_p = model.state_parameter_name(d)
+                    t = cs.DM.size(solution[n_s_x])[1]
+                    solution[n_s] = cs.DM(model.dim, t)
+                    solution[n_s][model.optimized_joint_indexes, :] = solution[n_s_x]
+                    solution[n_s][model.parameter_joint_indexes, :] = self._p_dict[
```

Comment:
This can fail when `n_s_p` does not appear in `_p_dict`. The user can choose to not provide all the parameters, and in this case they are assumed to be zero. The way to get round this is as follows.
```python
solution[n_s][model.parameter_joint_indexes, :] = self._p_dict.get(n_s_p, default)
```
where `default = cs.DM.zeros(model.dim, model.T)`.

**NOTE**: this will need a small modification to the `OptimizationBuilder` class in the `__init__` method. I'll make a note there too.

Reply:
I solved this problem.
Check my answer above. And the solution seems to me to be cleaner.

Reply:
I think what you do above will fail. Perhaps make this change? And revert above.

Reply:
No, it doesn't fail.
The reset_parameters() makes sure that they are initialised as you were doing before.
It is just doing the same thing as before.

Reply:
Ok, all good.